File: 3302-Find the Lexicographically Smallest Valid Sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> List[int]:
        N1, N2 = len(word1), len(word2)
        last = [-1]*N2

        idx2 = N2-1
        for i in range(N1-1, -1, -1):
            if word1[i] == word2[idx2]:
                last[idx2] = i
                idx2 -= 1
            if idx2 < 0:
                break
        
        wild = 1
        idx2 = 0
        res = []
        for i in range(N1):
            if idx2 == N2:
                break
            if word1[i] == word2[idx2]:
                res.append(i)
                idx2 += 1
            elif wild == 1 and (idx2 == N2-1 or i+1 <= last[idx2+1]):
                wild -= 1
                res.append(i)
                idx2 += 1
        
        return res if len(res) == N2 else []

File: C++/3302-Find the Lexicographically Smallest Valid Sequence.cpp
class Solution {
public:
    vector<int> validSequence(string word1, string word2) {
        int sz1 = word1.size(), sz2 = word2.size();
        vector<int> dp(sz1+1, 0);  // the number of characters word1[i:] matches word2


        // suffix preprocessing
        int idx2 = sz2-1;
        for (int i = sz1-1; i>=0; i--) {
            if (idx2 >= 0 && word1[i] == word2[idx2]) {
                idx2--;
                dp[i] = dp[i+1] + 1;
            } else {
                dp[i] = dp[i+1];
            }
        }

        // greedy
        idx2 = 0;
        int wild = 1;
        vector<int> res;
        for (int i=0; i<sz1; i++) {
            if (idx2 == sz2) break;
            if (word1[i] == word2[idx2]) {
                res.push_back(i);
                idx2++;
            } else if (wild == 1 && res.size() + wild + dp[i+1] >= sz2) {
                res.push_back(i);
                idx2++;
                wild--;
            }
        }

        if (res.size() < sz2) {
            return vector<int>{};
        }

        return res;
    }
};